src/SFA.DAS.EAS.Portal.Client/PortalClient.cs

```diff
@@ -1,23 +1,58 @@
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using SFA.DAS.EAS.Portal.Client.Application.Queries;
+using SFA.DAS.EAS.Portal.Client.Services.DasRecruit;
 using SFA.DAS.EAS.Portal.Client.Types;
 using StructureMap;
 
 namespace SFA.DAS.EAS.Portal.Client
 {
     public class PortalClient : IPortalClient
     {
-        private readonly GetAccountQuery _getAccountQuery;
-
+        private readonly IGetAccountQuery _getAccountQuery;
+        private readonly IDasRecruitService _dasRecruitService;
+        
         public PortalClient(IContainer container)
         {
-            _getAccountQuery = container.GetInstance<GetAccountQuery>();
+            _getAccountQuery = container.GetInstance<IGetAccountQuery>();
+            _dasRecruitService = container.GetInstance<IDasRecruitService>();
         }
         
-        public Task<Account> GetAccount(long accountId, CancellationToken cancellationToken = default)
+        //todo: might be better to just accept publicHashedAccountId, and decode it here
+        public async Task<Account> GetAccount(long accountId, string publicHashedAccountId,
+            bool hasPayeScheme, CancellationToken cancellationToken = default)
         {
-            return _getAccountQuery.Get(accountId, cancellationToken);
+            // we potentially map 1 more vacancy than necessary, but it keeps the code clean
+            var vacanciesTask = hasPayeScheme ? 
+                _dasRecruitService.GetVacancies(publicHashedAccountId, 2, cancellationToken) : null;
+
+            var account = await _getAccountQuery.Get(accountId, cancellationToken);
+
+            // at a later date, we might want to create an empty account doc and add the vacancy details to it, but for now, let's keep it simple
+            if (!hasPayeScheme || account == null)
+                return account;
+            
+            var vacancies = await vacanciesTask;
+            if (vacancies == null)
+                return account;
+            
+            var vacanciesCount = vacancies.Count();
+            switch (vacanciesCount)
+            {
+                case 0:
+                    account.VacancyCardinality = Cardinality.None;
+                    break;
+                case 1:
+                    account.VacancyCardinality = Cardinality.One;
+                    account.SingleVacancy = vacancies.Single();
```

Comment:
I think this is a worthwhile optimisation (and improves the interface). Previously, this PR requested up to a 1000 vacancy records, mapped them all, then returned them all to the consumer, when all the consumer needs to know is if there is 0, 1 or >1 vacancies, and the details of a single vacancy, if there is only one.

What it now does is return an enum stating f there are 0, 1 or >1 vacancies and the details of a maximum of 1 vacancy.

This means that instead of the vacancy API having to read up to 1000 full records from its local store, return those records over the network, then the client storing them in memory, mapping them to the required subset of properties, then passing them to the consumer (and then all that memory having to be garbage collected), the vacancy API only has to read and send a maximum of two records, the client only maps a maximum of two records (could be 1!) and then returns a maximum of 1 record subset.

This should have a considerable impact on the time taken to render the homepage, and the reduction in memory usage and churn should also enable the website and vacancy API to perform its other actions quicker.

It also makes the interface more intuitive, as it wasn't obvious before that it could be returning only a subset of the entire vacancy set, whereas now it's more obvious what is being returned.

If we wanted to to return the full set of vacancies, we'd have to write extra code to page through the vacancies returned from the API, adding unnecessary complexity.

It's also easy to change the interface if requirements change. The vacancy records are retrieved from the API each time and added to the account fetched from the document collection. We don't store the vacancies in the account document in the CosmosDB document collection, so we don't have to worry about backwards compatibility while reading the account, or updating existing account documents if the schema changes.

If requirements change, we can change the interface to match the new requirements. We might, for example, want to display a maximum of 5 records in a tabbed interface, but I doubt if we'd ever want to display hundreds of vacancies on the homepage, at least not without paging, so I doubt if we'll ever need to return the full set of vacancies.

Reply:
This is implementing a UI requirement in the back end rather than supporting the requirements of the UI.

I agree with Corey that the model returned shouldn't have this UI rule in it.  The original model had a collection of Vacancies in it which meets the requirements.  The cardinality is implicit by the number of records in that collection so we don't need this VacancyCardinality property on the account model.  As far as the back end is concerned (being a service) the UI requires all the records back for that account (since it may be populating a drop down list so the required vacancy can be selected).  If there is only one vacancy record then only one will be returned so it performs for that scenario.  This is not an optimisation but a corruption of the model, optimisations should only be done when there is a requirement for them based on empirical measurements.  A better approach to optimise would be to include in the Filter object how many records to return and a value specifying if there are more records to return.  i.e. Filter.MaxVacancy = -1 .  So then the client can decide if an optimisation is required since it can set the value to -1 (All), 1, 3 or 4 or to what ever the UI requirement is and therefore is no longer constrained by the back end implementation.  In this case a count of the total number of records would be useful to return.  By taking this approach if the UI requirements change then the back end can support those requirements without being changed.  Likewise a full or summary vacancy resultset can be returned by setting another filter property.

Reply:
I like the idea of extending the filter and adding the max vacancies property. I think this retains the optimisation that Phil has introduced without changing the account schema based on current requirements. This would mean there would be little change required should the consumer of the client or the source of the Vacancy data change.

Reply:
I have addressed your concerns in the order that you raised them...

Yes, if the client returned the full set of Vacancies, then the `VacancyCardinality` property would not be needed, but by not returning the full set, it definitely is.

The front end does not require all the records back, as proven by the requirements being met by the client only returning a maximum of 1 record. The UI is *not* populating a select list, and I think it extremely unlikely that we'll ever be asked to populate a select list with up to a thousand options. If we *were* ever tasked to implement that, then we are free to change the code to match any new requirements.

Yes, if there is only one vacancy, then performance will be fine. It was not that scenario that I was worried would slow down the rendering of the homepage.

I agree in principle with not optimising prematurely, and optimizing based on empirical measurements, but common sense has to prevail and we should never fetch a thousand records when we only need one. Just because you shouldn't optimise prematurely, doesn't mean that you should make the system do extra unnecessary work, and I don't need to measure results to know that fetching two records is going to be considerably quicker than fetching a 1000!

With regards to your suggestion of passing a filter to the client, that would involve implementing paging in the service that talks to the vacancy API, and implementing paging in the client itself, all for some potential future requirement that may never come. We should keep the code as simple as possible to implement the current requirements, and if new requirements come along, we can adapt the system to best fit the *actual* new requirements, rather than some hypothetical requirements. A new requirement may be to replace the content of the panel currently radiating vacancy information, with information from another system. In that case, implementing two sets of paging code would have been a waste of time and money.

I have the same issue with your suggestion of passing a flag to the client to get it to return either a full or summary vacancy record. There is no requirement to return a full vacancy record, so there is no point in spending time implementing such functionality. That road leads to an unnecessarily complex system because future requirements have been guessed at. The unnecessary complexity also makes it harder to refactor the system for actual new requirements.

The considerations would be different if we were implementing a general library or framework, where you have to guess at consumer requirements, but we're effectively implementing a service for a single consumer, where both are released together.

Reply:
We could apply a page size of whatever the filter max vacancies property is set to which would avoid the need to implement paging logic. This way we would get the benefits mentioned in my previous comment. Setting the value to 2 in the caller would give us all the information we need currently from vacancies. If a time came when a consumer asks for a thousand vacancies we would optimise and add paging at that point.

This I think would be a good compromise with fairly minimal changes.

Reply:
So basically, you're suggesting to move the logic about how many vacancies to fetch, and the cardinality calculation, out of the client and up into the website. (That would effectively mean moving roughly half of the client's `GetAccount` method into the website.)

So the debate has basically boiled down to where the logic associated with the portal should live. 

I personally think it's cleaner to encapsulate such knowledge in the portal's code, rather than having it in the controller/website, (we already have domain knowledge in the portal), but if you want the logic moved into the website, I can move it.

(Having the logic in the website also has the disadvantage of returning more data than is required, but only up to twice the number of vacancies, rather than a 1000 😄.)

Reply:
I think it makes more sense for the website (consumer of the client) to decide how to use the data that it is presented. The logic itself would be encapulated within the viewmodel then as it derives the cardinality and therefore which view to display from the returned vacancies.

Reply:
The portal is a read only model - the client is just allowing a client to query that data.  If the portal received the vacancy data via events would you design the interface in this way?  Would you still pass a HasPayeScheme in?  You are leaking implementation details rather than encapsulating them.

Reply:
Deciding how many vacancies to fetch and counting the vacancies, has been moved out of the portal client (and the related website PR's will be refactored to include that code), and a parameter object has been introduced to pass the business parameters to the `GetAccount` method (as agree offline).